OpenSim/Common/Component.cpp

```diff
@@ -1526,11 +1529,39 @@ void Component::dumpConnections() const {
 
 
 void Component::initComponentTreeTraversal(const Component &root) const {
-    // Going down the tree, node is followed by all its
-    // children in order, last child's successor is the parent's successor.
+    // Going down the tree, this node is followed by all its children.
+    // The last child's successor (next) is the parent's successor.
+
+    const size_t nmsc = _memberSubcomponents.size();
+    const size_t npsc = _propertySubcomponents.size();
+    const size_t nasc = _adoptedSubcomponents.size();
+
+    // If this isn't the root component and it has no parent, then
+    // this is an orphan component and we likely failed to call 
+    // finalizeFromProperties on the root OR this is a clone that
+    // has not been added to the root (in which case would have a parent).
+    if ((this != &root) && !this->hasParent() ) {
+        OPENSIM_THROW(ComponentIsAnOrphan, getName(), getConcreteClassName());
+    }
+
+    if ((this == &root) && !(nmsc + npsc + nasc)) {
+        ComponentIsRootWithNoSubcomponents ex(__FILE__, __LINE__, __func__,
+            getName(), getConcreteClassName());
+        if (getConcreteClassName() == "Model") {
```

Comment:
1. I agree it is ugly, but Model was the only case I could be certain about.
2. I definitely can do (and contemplated doing) that and I even used a `try/catch` in an previous incarnation of `testComponents`. I opted not to do that because I thought throwing an exception when accessing an empty component maybe too severe. But, you raise a good point that it indicates that the user may not realize that they are mistakenly dealing with an empty component. I am happy to make it more strict. It does make it more difficult to write generic code that takes a Component and rummages through it. That might be something the GUI does routinely, but even then it would be strange to be to have a Component that had no parent and was empty and might be worth alerting about. @aymanhab what do you think?